casper/src/main/scala/coop/rchain/casper/Estimator.scala

```diff
@@ -50,83 +52,44 @@ object Estimator {
         IndexedSeq(b)
       }
     }
+
     def stillSame(blocks: IndexedSeq[BlockHash], newBlocks: IndexedSeq[BlockHash]) =
       newBlocks == blocks
 
-    val scoresMap = buildScoresMap(blockDag, internalMap)
-    sortChildren(IndexedSeq(genesis.blockHash), blockDag.childMap, scoresMap)
-      .map(internalMap)
+    for {
+      scoresMap           <- buildScoresMap[F](blockDag)
+      sortedChildrenHash  = sortChildren(IndexedSeq(genesis.blockHash), blockDag.childMap, scoresMap)
+      maybeSortedChildren <- sortedChildrenHash.toList.traverse(BlockStore[F].get)
+      sortedChildren      = maybeSortedChildren.flatten.toVector
+    } yield sortedChildren
   }
 
-  def buildScoresMap(blockDag: BlockDag,
-                     internalMap: Map[BlockHash, BlockMessage]): Map[BlockHash, Int] = {
-    def hashParents(internalMap: Map[BlockHash, BlockMessage],
-                    hash: BlockHash): Iterator[BlockHash] = {
-      val b = internalMap(hash)
-      parents(b).iterator
-    }
+  def buildScoresMap[F[_]: Monad: BlockStore](blockDag: BlockDag): F[Map[BlockHash, Int]] = {
+    def hashParents(hash: BlockHash): F[List[BlockHash]] =
+      for {
+        b <- getBlock[F](hash)
+      } yield parents(b).toList
 
     def addValidatorWeightDownSupportingChain(scoreMap: Map[BlockHash, Int],
-                                              internalMap: Map[BlockHash, BlockMessage],
                                               validator: Validator,
-                                              latestBlockHash: BlockHash) =
-      DagOperations
-        .bfTraverse[BlockHash](Some(latestBlockHash))(hashParents(internalMap, _))
-        .foldLeft(scoreMap) {
-          case (acc, hash) =>
-            val b         = internalMap(hash)
-            val currScore = acc.getOrElse(hash, 0)
-
-            val validatorWeight = weightFromValidator(b, validator, internalMap)
-
-            acc.updated(hash, currScore + validatorWeight)
-        }
+                                              latestBlockHash: BlockHash): F[Map[BlockHash, Int]] =
+      for {
+        dag <- DagOperations.bfTraverseF[F, BlockHash](List(latestBlockHash))(hashParents).toList
```

Comment:
We need a Foldable[StreamT] instance apparently. I'm not sure how to obtain that.

Reply:
I defined a `foldLeft` function on `StreamT` directly, so you should be able to just call that. https://github.com/rchain/rchain/blob/dev/shared/src/main/scala/coop/rchain/shared/StreamT.scala#L61

Reply:
Hmm I think I need a

`def foldLeft[B](b: B)(f: (B, A) => F[B])(implicit monad: Monad[F]): F[B]`

instead of

`def foldLeft[B](b: B)(f: (B, A) => B)(implicit monad: Monad[F]): F[B]`

Reply:
Does this work

```
  def foldLeftF[B](b: B)(f: (B, A) => F[B])(implicit monad: Monad[F]): F[B] = self match {
    case SCons(curr, lazyTail) =>
      for {
        newB <- f(b, curr)
        result <- lazyTail.value.flatMap(_.foldLeftF(newB)(f))
      } yield result

    case SLazy(lazyTail) => lazyTail.value.flatMap(_.foldLeftF(b)(f))

    case _: SNil[F] => b.pure[F]
  }
```

? I've noticed that I usually don't want to do .pure[F] in the "initAcc" spot.